src/Oscoin/Protocol.hs

```diff
@@ -0,0 +1,275 @@
+                      let evt = BlockValidationFailedEvent (blockHash incomingBlock) validationError
+                      in pure (mgr, [evt])
+                  Right () -> do
+                      -- Step 3: store the fully-validated block.
+                      BlockStore.insertBlock bsPrivateAPI incomingBlock
+                      pure (mgr, mempty)
+
+          -- NOTE(adn) This is a side effect of the fact that storing
+          -- the genesis as an orphan would insert it into the orphanage with
+          -- its parent hash, which means that we won't be able to retrieve
+          -- a potentiol candidate branching off from genesis in our 'selectBestChain'
+          -- function (cfr. QuickCheck seed 554786). To avoid that, we avoid
+          -- inserting the genesis block into the orphanage in the first place;
+          -- After all, the genesis block should never be considered at orphan.
+          else map (,mempty) . pure $
+              if not (isGenesisBlock incomingBlock)
+                  then mgr { protoOrphanage = O.insertOrphan incomingBlock (protoOrphanage mgr) }
+                  else mgr
+
+    -- Step 4: Chain selection.
+    (,evts) <$> selectBestChain mgr'
+  where
+      bsPublicAPI :: BlockStoreReader c tx s m
+      bsPublicAPI = fst . protoFullBlockStore $ mgr
+
+      bsPrivateAPI :: BlockStoreWriter c tx s m
+      bsPrivateAPI = snd . protoFullBlockStore $ mgr
+
+{------------------------------------------------------------------------------
+  (Best) Chain selection
+------------------------------------------------------------------------------}
+
+-- | Performs best chain selection and returns the new 'Protocol' data
+-- structure.
+selectBestChain
+    :: forall m s c tx.
+       ( Monad m
+       , Ord s
+       , Ord (BlockHash c)
+       )
+    => Protocol c tx s m
+    -> m (Protocol c tx s m)
+selectBestChain mgr = do
+    -- NOTE(adn) Grab the \"mutable\" part of the chain and
+    -- compare each root hash to see if there is any fork originating from that.
+    let depth = fromIntegral $ mutableChainDepth (protoConfig mgr)
+    mutableBlockHashes <-
+        map blockHash . Chrono.reverse <$> BlockStore.getBlocksByDepth bsPublicAPI depth
+
+    case O.selectBestChain (toOldestFirst mutableBlockHashes) orphanage of
+        Just (parentHash, bestChain) | isValid bestChain -> do
+            -- Compare the orphan best chain with the chain suffix currently adopted
+            chainSuffix <- BlockStore.getBlocksByParentHash bsPublicAPI parentHash
+
+            -- Switch-to-better-chain condition: either the chain suffix is
+            -- empty, which means we are extending directly the tip, or if we
+            -- found a better candidate.
+            let scoreFn       = protoScoreBlock mgr
+                newChainFound =
+                    null chainSuffix
+                 || O.fromChainSuffix scoreFn (NonEmpty.fromList . toNewestFirst $ chainSuffix) < bestChain
```

Comment:
This is suspicious to me, why did we find a new chain if it is lesser than the best chain?

Reply:
oh, because the `Ord` instance compares by block score (and height!) so this is essentially saying that if the chain suffix _of the main chain_ is "less" than the best chain, we switch.

Reply:
Ah, I understand your confusion now -- no no, the naming perhaps doesn't help -- `bestChain` is actually the chain _we extracted from the orphanage_, not the current, adopted one 😉

Reply:
Oh! yes it's the naming. Maybe even reversing the condition would help:
`bestChain > O.fromChainSuffix ...`, or renaming `bestChain` to `bestFork`